File: design_patterns/10_State/Order.java
// OrderState.java
public interface OrderState {
    void nextState(Order order);
    void prevState(Order order);
    String toString();
}

// NewOrderState.java
public class NewOrderState implements OrderState {
    @Override
    public void nextState(Order order) {
        order.setState(new ProcessingOrderState());
    }

    @Override
    public void prevState(Order order) {
        System.out.println("The order is already in the initial state.");
    }

    @Override
    public String toString() {
        return "New Order";
    }
}

// ProcessingOrderState.java
public class ProcessingOrderState implements OrderState {
    @Override
    public void nextState(Order order) {
        order.setState(new ShippedOrderState());
    }

    @Override
    public void prevState(Order order) {
        order.setState(new NewOrderState());
    }

    @Override
    public String toString() {
        return "Processing Order";
    }
}

// ShippedOrderState.java
public class ShippedOrderState implements OrderState {
    @Override
    public void nextState(Order order) {
        order.setState(new DeliveredOrderState());
    }

    @Override
    public void prevState(Order order) {
        order.setState(new ProcessingOrderState());
    }

    @Override
    public String toString() {
        return "Shipped Order";
    }
}

// DeliveredOrderState.java
public class DeliveredOrderState implements OrderState {
    @Override
    public void nextState(Order order) {
        System.out.println("The order is already in the final state.");
    }

    @Override
    public void prevState(Order order) {
        order.setState(new ShippedOrderState());
    }

    @Override
    public String toString() {
        return "Delivered Order";
    }
}

// Order.java
public class Order {
    private OrderState state;

    public Order() {
        state = new NewOrderState();
    }

    public void setState(OrderState state) {
        this.state = state;
    }

    public void nextState() {
        state.nextState(this);
    }

    public void prevState() {
        state.prevState(this);
    }

    @Override
    public String toString() {
        return "Order is currently in '" + state + "' state.";
    }

    public static void main(String[] args) {
        Order order = new Order();
        System.out.println(order);  // Output: Order is currently in 'New Order' state.

        order.nextState();
        System.out.println(order);  // Output: Order is currently in 'Processing Order' state.

        order.nextState();
        System.out.println(order);  // Output: Order is currently in 'Shipped Order' state.

        order.nextState();
        System.out.println(order);  // Output: Order is currently in 'Delivered Order' state.

        order.prevState();
        System.out.println(order);  // Output: Order is currently in 'Shipped Order' state.

        order.prevState();
        System.out.println(order);  // Output: Order is currently in 'Processing Order' state.

        order.prevState();
        System.out.println(order);  // Output: Order is currently in 'New Order' state.
    }
}

File: design_patterns/10_State/order.js
class OrderState {
    nextState(order) {
        throw new Error('nextState method should be implemented');
    }
    
    prevState(order) {
        throw new Error('prevState method should be implemented');
    }

    toString() {
        throw new Error('toString method should be implemented');
    }
}

class NewOrderState extends OrderState {
    nextState(order) {
        order.state = new ProcessingOrderState();
    }

    prevState(order) {
        console.log("The order is already in the initial state.");
    }

    toString() {
        return 'New Order';
    }
}

class ProcessingOrderState extends OrderState {
    nextState(order) {
        order.state = new ShippedOrderState();
    }

    prevState(order) {
        order.state = new NewOrderState();
    }

    toString() {
        return 'Processing Order';
    }
}

class ShippedOrderState extends OrderState {
    nextState(order) {
        order.state = new DeliveredOrderState();
    }

    prevState(order) {
        order.state = new ProcessingOrderState();
    }

    toString() {
        return 'Shipped Order';
    }
}

class DeliveredOrderState extends OrderState {
    nextState(order) {
        console.log("The order is already in the final state.");
    }

    prevState(order) {
        order.state = new ShippedOrderState();
    }

    toString() {
        return 'Delivered Order';
    }
}

class Order {
    constructor() {
        this.state = new NewOrderState();
    }
    
    nextState() {
        this.state.nextState(this);
    }
    
    prevState() {
        this.state.prevState(this);
    }
    
    toString() {
        return `Order is currently in '${this.state.toString()}' state.`;
    }
}

// Example Usage
const order = new Order();
console.log(order.toString());  // Output: Order is currently in 'New Order' state.

order.nextState();
console.log(order.toString());  // Output: Order is currently in 'Processing Order' state.

order.nextState();
console.log(order.toString());  // Output: Order is currently in 'Shipped Order' state.

order.nextState();
console.log(order.toString());  // Output: Order is currently in 'Delivered Order' state.

order.prevState();
console.log(order.toString());  // Output: Order is currently in 'Shipped Order' state.

order.prevState();
console.log(order.toString());  // Output: Order is currently in 'Processing Order' state.

order.prevState();
console.log(order.toString());  // Output: Order is currently in 'New Order' state.
